File: rxfirebase2-database/src/main/java/rxfirebase2/database/RxFirebaseDatabase.java
package rxfirebase2.database;

import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.GenericTypeIndicator;
import com.google.firebase.database.MutableData;
import com.google.firebase.database.Query;
import com.google.firebase.database.Transaction;

import io.reactivex.annotations.CheckReturnValue;
import io.reactivex.annotations.NonNull;
import rxfirebase2.database.model.DataValue;
import rxfirebase2.database.transformers.SingleTransformerOfClazz;
import rxfirebase2.database.transformers.SingleTransformerOfGenericTypeIndicator;
import rxfirebase2.database.transformers.TransformerOfClazz;
import rxfirebase2.database.transformers.TransformerOfGenericTypeIndicator;

import java.util.Map;
import java.util.concurrent.Callable;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.functions.Function;
import rxtasks2.RxTask;


public final class RxFirebaseDatabase {

    /**
     * @param query
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static Observable<ChildEvent> childEvents(@NonNull final Query query) {
        return Observable.create(new QueryChildEventsOnSubscribe(query));
    }

    /**
     * @param query
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static Single<DataSnapshot> data(@NonNull final Query query) {
        return RxValue.single(query);
    }

    /**
     * @param query
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static Observable<DataSnapshot> dataChanges(@NonNull final Query query) {
        return RxValue.changes(query);
    }

    /**
     * @param ref
     * @param clazz
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Observable<DataValue<T>> dataChangesOf(
            @NonNull final DatabaseReference ref, @NonNull final Class<T> clazz) {
        return dataChanges(ref).compose(new TransformerOfClazz<>(clazz));
    }

    /**
     * @param query
     * @param clazz
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Observable<DataValue<T>> dataChangesOf(
            @NonNull final Query query, @NonNull final Class<T> clazz) {
        return dataChanges(query).compose(new TransformerOfClazz<>(clazz));
    }

    /**
     * @param ref
     * @param typeIndicator
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Observable<DataValue<T>> dataChangesOf(
            @NonNull DatabaseReference ref, @NonNull GenericTypeIndicator<T> typeIndicator) {
        return dataChanges(ref)
                .compose(new TransformerOfGenericTypeIndicator<T>(typeIndicator));
    }

    /**
     * @param query
     * @param typeIndicator
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Observable<DataValue<T>> dataChangesOf(
            @NonNull Query query, @NonNull GenericTypeIndicator<T> typeIndicator) {
        return dataChanges(query)
                .compose(new TransformerOfGenericTypeIndicator<T>(typeIndicator));
    }

    /**
     * @param ref
     * @param clazz
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Single<T> dataOf(
            @NonNull DatabaseReference ref, @NonNull Class<T> clazz) {
        return data(ref).compose(new SingleTransformerOfClazz<>(clazz));
    }

    /**
     * @param query
     * @param clazz
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Single<T> dataOf(
            @NonNull Query query, @NonNull Class<T> clazz) {
        return data(query).compose(new SingleTransformerOfClazz<T>(clazz));
    }

    /**
     * @param ref
     * @param typeIndicator
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Single<T> dataOf(
            @NonNull DatabaseReference ref, @NonNull GenericTypeIndicator<T> typeIndicator) {
        return data(ref).compose(new SingleTransformerOfGenericTypeIndicator<T>(typeIndicator));
    }

    /**
     * @param query
     * @param typeIndicator
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Single<T> dataOf(
            @NonNull Query query, @NonNull GenericTypeIndicator<T> typeIndicator) {
        return data(query).compose(new SingleTransformerOfGenericTypeIndicator<T>(typeIndicator));
    }

    /**
     * @param ref
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static Completable removeValue(@NonNull final DatabaseReference ref) {
        return RxTask.completes(new Callable<Task<Void>>() {
            @Override
            public Task<Void> call() throws Exception {
                return ref.removeValue();
            }
        });
    }

    /**
     * @param ref
     * @param priority
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static Completable setPriority(
            @NonNull final DatabaseReference ref, @NonNull final Object priority) {
        return RxTask.completes(new Callable<Task<Void>>() {
            @Override
            public Task<Void> call() throws Exception {
                return ref.setPriority(priority);
            }
        });
    }

    /**
     * @param ref
     * @param value
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Completable setValue(
            @NonNull final DatabaseReference ref,
            @NonNull final T value) {
        return RxTask.completes(new Callable<Task<Void>>() {
            @Override
            public Task<Void> call() throws Exception {
                return ref.setValue(value);
            }
        });
    }

    /**
     * @param ref
     * @param value
     * @param priority
     * @param <T>
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static <T> Completable setValue(
            @NonNull final DatabaseReference ref,
            @NonNull final T value,
            @NonNull final Object priority) {
        return RxTask.completes(new Callable<Task<Void>>() {
            @Override
            public Task<Void> call() throws Exception {
                return ref.setValue(value, priority);
            }
        });
    }

    /**
     * @param ref
     * @param function
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static Single<DataSnapshot> runTransaction(
            @NonNull final DatabaseReference ref,
            @NonNull final Function<MutableData, Transaction.Result> function) {
        return RxValue.transaction(ref, function);
    }

    /**
     * @param ref
     * @param fireLocalEvents
     * @param function
     * @return
     * ref.runTransaction(handler, fireLocalEvents);
     */
    @NonNull
    @CheckReturnValue
    public static Single<DataSnapshot> runTransaction(
            @NonNull final DatabaseReference ref,
            @NonNull final Function<MutableData, Transaction.Result> function,
            final boolean fireLocalEvents) {
        return RxValue.transaction(ref, function, fireLocalEvents);
    }

    /**
     * @param ref
     * @param map
     * @return
     */
    @NonNull
    @CheckReturnValue
    public static Completable updateChildren(
            @NonNull final DatabaseReference ref, @NonNull final Map<String, Object> map) {
        return RxTask.completes(new Callable<Task<Void>>() {
            @Override
            public Task<Void> call() throws Exception {
                return ref.updateChildren(map);
            }
        });
    }

    private RxFirebaseDatabase() {
        throw new AssertionError("No instances");
    }
}

File: rxfirebase2-database-kotlin/src/main/kotlin/rxfirebase2/database/RxFirebaseDatabase.kt
@file:Suppress("NOTHING_TO_INLINE")

package rxfirebase2.database

import rxfirebase2.database.model.DataValue
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.GenericTypeIndicator
import com.google.firebase.database.MutableData
import com.google.firebase.database.Query
import com.google.firebase.database.Transaction
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single

inline fun DatabaseReference.childEvents()
        : Observable<ChildEvent>
        = RxFirebaseDatabase.childEvents(this)

inline fun DatabaseReference.data()
        : Single<DataSnapshot>
        = RxFirebaseDatabase.data(this)

inline fun DatabaseReference.dataChanges()
        : Observable<DataSnapshot>
        = RxFirebaseDatabase.dataChanges(this)

inline fun <reified T : Any> DatabaseReference.dataChangesOf()
        : Observable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, T::class.java)

inline fun <reified T : Any> DatabaseReference.dataChangesOf(typeIndicator: GenericTypeIndicator<T>)
        : Observable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, typeIndicator)

inline fun <reified T : Any> DatabaseReference.dataOf()
        : Single<T>
        = RxFirebaseDatabase.dataOf(this, T::class.java)

inline fun <reified T : Any> DatabaseReference.dataOf(typeIndicator: GenericTypeIndicator<T>)
        : Single<T>
        = RxFirebaseDatabase.dataOf(this, typeIndicator)

inline fun DatabaseReference.removesValue()
        : Completable
        = RxFirebaseDatabase.removeValue(this)

inline fun DatabaseReference.runsTransaction(noinline task: (MutableData) -> Transaction.Result)
        : Completable
        = RxFirebaseDatabase.runTransaction(this, task)

inline fun DatabaseReference.runsTransaction(
        fireLocalEvents: Boolean, noinline task: (MutableData) -> Transaction.Result)
        : Completable
        = RxFirebaseDatabase.runTransaction(this, fireLocalEvents, task)

inline fun DatabaseReference.setsPriority(priority: Any)
        : Completable
        = RxFirebaseDatabase.setPriority(this, priority)

inline fun <reified T : Any> DatabaseReference.setsValue(value: T)
        : Completable
        = RxFirebaseDatabase.setValue(this, value)

inline fun <reified T : Any> DatabaseReference.setsValue(value: T, priority: Any)
        : Completable
        = RxFirebaseDatabase.setValue(this, value, priority)

inline fun DatabaseReference.updatesChildren(update: Map<String, Any?>)
        : Completable
        = RxFirebaseDatabase.updateChildren(this, update)

inline fun Query.childEvents()
        : Observable<ChildEvent>
        = RxFirebaseDatabase.childEvents(this)

inline fun Query.data()
        : Single<DataSnapshot>
        = RxFirebaseDatabase.data(this)

inline fun Query.dataChanges()
        : Observable<DataSnapshot>
        = RxFirebaseDatabase.dataChanges(this)

inline fun <reified T : Any> Query.dataChangesOf()
        : Observable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, T::class.java)

inline fun <reified T : Any> Query.dataChangesOf(typeIndicator: GenericTypeIndicator<T>)
        : Observable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, typeIndicator)

inline fun <reified T : Any> Query.dataOf()
        : Single<T>
        = RxFirebaseDatabase.dataOf(this, T::class.java)

inline fun <reified T : Any> Query.dataOf(typeIndicator: GenericTypeIndicator<T>)
        : Single<T>
        = RxFirebaseDatabase.dataOf(this, typeIndicator)
